Rfc nightly build (#6555)
Scille/parsec-cloud: docs/rfcs/1008-nightly-build.md

#### Include the client part in the releaser workflow

We need to add the client to the workflow [`releaser.yml`].
Thankfully we already have the workflow for that ([`package-ionic-app.yml`]).

[`package-ionic-app.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/package-server.yml
[`releaser.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/releaser.yml

### Overwriting nightly release & automatic release

Currently, the [`releaser.yml`] workflow create a [GitHub release] in draft mode.

[GitHub release]: https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases

This is problematic if the workflow create a `draft-release` for each nightly build (everyday so).

> We would be overwhelmed by the amount of draft releases that would accumulate.

To circumvent that we could make the workflow to remove the previous nightly release.
But we have some limitation, the API to [delete a release] require a `release_id`.

> [delete a release]: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#delete-a-release

Author: We could do that. I've choose that approach of re-creating the release because GitHub sort the releases by creation date. that mean the nightly release will be put behind older releases.

> [delete a release]: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#delete-a-release

Reviewer: You mean it will be put behind _newer_ releases? That could be inconvenient later, but I think it's no big deal for the moment. Right now, there are no other release planned on the `v2.x` or `v3.x` branch and we need a nightly release _of the upcoming v3.0_. We could improve the nigthly workflow later, based on this RFC.
WDYT?

> [delete a release]: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#delete-a-release

Reviewer: Maybe we can put the two alternatives (update vs delete+create) in the RFC, just to document them so we can merge the PR.

> [delete a release]: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#delete-a-release

Author: For tracking purpose, we have a paragraph that mention `override the assets`

This ID could be obtained by [getting a release using its tag name] but a `draft-release` isn't identified by a tag name (the more you know)

[getting a release using its tag name]: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-a-release-by-tag-name

> [!TIP]
> This could explain why `Github CLI` only delete releases identified by a tag only, see [`gh release delete` manual](https://cli.github.com/manual/gh_release_delete).

> We could also [list all releases] to find the one we look for.

[list all releases]: https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#list-releases

So we could update the workflow to directly create the release.

> At this step, you may be asking yourself [Why do we need to release the nightly build ?](#why-do-we-need-to-release-the-nightly-build-or-why-we-cant-directly-use-the-artifact-link-of-a-workflow)

> ### Snapcraft channel with tracks

Author: You mean that we should push the nightly build into latest ?

> ### Snapcraft channel with tracks

Reviewer: Resolving the conversation as this is no longer relevant (see https://github.com/Scille/parsec-cloud/pull/6590)

### Schedule the workflow

Require [Overwriting nightly release & automatic release](#overwriting-nightly-release--automatic-release)

We need to have a workflow that schedule the nightly build at a specific interval (every day at 22H30 for example).
GitHub already allow to [trigger a workflow on a schedule].

[trigger a workflow on a schedule]: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule

> [!CAUTION]
> Event triggered by a GitHub actions don't trigger a workflow. So everything should be done in a single workflow.

### Slack notification (optional)

We could send a Slack notification when a nightly build finished (successfully or not)
With the link to download the generated build.

Slack provide GitHub action for that: <https://github.com/slackapi/slack-github-action>

### Improve the unique dev format

Currently, `releaser.py` generate a dev version by appending the `dev` part at the end of the current version before the local part.

The dev part is formatted that way `dev.{number}`. The `{number}` value is calculated from the current date with the following formula:

> $f(year, month, day) = year * 366 + month * 31 + day$

Reviewer: Where do this formula come from?

It is not clear to me why is this an _improvement_ (as in the section title)? Is there a particular problem/issue with the current formula (besides the mistake on year factor)? 
If the answer is no, is probably better to open an issue to "fix" the dev format, proposing the alternatives that you listed here.

Author: > Where do this formula come from?

From my hat :tophat:

> It is not clear to me why is this an improvement ...

For me some formulas are simpler to use

> Is there a particular problem/issue with the current formula (besides the mistake on year factor)?

The problem is that value overlap between December and January:

```math
\begin{align}
f(y, m, d) = y * 366 + m * 31 + d \\
f(0, 12, 26) = f(1, 1, 1) = 398 \\
f(0, 12, 31) = f(1, 1, 6) = 403
\end{align}
```

Reviewer: > > Where do this formula come from?
> 
> From my hat 🎩

So indeed, better try to find a more standard way. I am not sure that we really need the date information, so maybe just an incremental index.

> > Is there a particular problem/issue with the current formula (besides the mistake on year factor)?
> 
> The problem is that value overlap between December and January:

IMO this is acceptable for the moment. I suggest dealing with this subject in a separate issue (so we can discuss there what is the best naming scheme).

Reviewer: Resolving the conversation as this is no longer relevant (see https://github.com/Scille/parsec-cloud/pull/6592)

- The corrected formulas using $372$:

  For `2024-02-27 15:50:28+01:00`, the value is calculated with:

  $$f'(year, month, day) = year * 372 + month * 31 + day$$
  $$f'(2024, 2, 27) = 753017$$

  That would give `dev.753017`

- A date (`dev.YYYYMMDD`): For `2024-02-27 15:50:28+01:00` That gives `dev.20240227`

## F.A.Q

### Why do we need to release the nightly build? Or why we can't directly use the artifact link of a workflow?

When a workflow generate artifacts, those artifacts could be directly downloaded from the workflow summary.

But that has some limitations:

- Those artifacts expire after a configured period (default to 90days).
- The artifacts are wrapped in a zip file, so you need to extract them before using them.
- You could not easily know if the workflow summary you're using is the latest available.
- The artifacts may contain data that is not useful for you.

### Why do we use the `dev` identifier for a nightly build ?

We need to have a version that is compatible with [semver] and [pep440], that indicate that it's a nightly build.

On [semver], we could use any verb like `nightly`, so the choice come from [pep440].

> Pep440 only allows the verb `dev` at the end of the version before the local identifiers ([Specify a development release in pep440])

Reviewer: I don't know if it makes sense, but how about adding `nigthly` to the local identifiers?

Author: The local identifier isn't used when comparing 2 version together

[semver]: https://semver.org/
[pep440]: https://peps.python.org/pep-0440/

> [Specify a development release in pep440]: https://peps.python.org/pep-0440/#developmental-releases

Reviewer: There is a note in PEP440 saying that the reference spec is in https://packaging.python.org/en/latest/specifications/version-specifiers/#developmental-releases

Author: They're virtually the same :shrug:

Reviewer: :man_shrugging:

  - [Why do we use the `dev` identifier for a nightly build ?](#why-do-we-use-the-dev-identifier-for-a-nightly-build-)

## Goals

- Periodically test the release pipeline.
- Have a recent build software at the ready (client and server).

## What we already have

We already have a workflow that create a release & build the artifact[^releaser-workflow] for that release.

That workflow[^releaser-workflow] is triggered when we modify the workflow itself but notably when we push a tag on the repository that have the format `v[0-9]+.[0-9].[0-9]+*`.

Currently, the workflow build the server python wheel for Linux, Windows & macOS.

[^releaser-workflow]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/releaser.yml

The script `releaser.py` can generate unique version that can be used for versioning and tagging nightly build.

## What is missing

From the release, we already have the server part indicated by:

<https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/releaser.yml#L74-L79>

> ### For the client part
> 
> #### Missing version argument in the workflow `package-ionic-app.yml`
> 
> The workflow ([`package-ionic-app.yml`]) need to be adapted like [`package-server.yml`] to take a version in argument.
> 
> [`package-server.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/package-server.yml
> 
> #### Include the client part in the releaser workflow
> 
> We need to add the client to the workflow [`releaser.yml`].
> Thankfully we already have the workflow for that ([`package-ionic-app.yml`]).
> 
> [`package-ionic-app.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/package-server.yml
> [`releaser.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/releaser.yml

Reviewer: This is quite verbose, I would summarize it:

```suggestion
### Client packaging

The following changes are required for the [`package-ionic-app.yml`] workflow:

- it needs to take a version argument (like [`package-server.yml`]).
- it needs to be added to the [`releaser.yml`] workflow.

[`package-server.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/package-server.yml
[`package-ionic-app.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/package-server.yml
[`releaser.yml`]: https://github.com/Scille/parsec-cloud/blob/05e4deba964c6b68b6fcbca4d458c235c43f6cff/.github/workflows/releaser.yml
```

### Overwriting nightly release & automatic release

> Currently, the [`releaser.yml`] workflow create a [GitHub release] in draft mode.
> 
> [GitHub release]: https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases
> 
> This is problematic if the workflow create a `draft-release` for each nightly build (everyday so).
> 
> > We would be overwhelmed by the amount of draft releases that would accumulate.
> 
> To circumvent that we could make the workflow to remove the previous nightly release.
> But we have some limitation, the API to [delete a release] require a `release_id`.

Reviewer: ```suggestion
The [`releaser.yml`] workflow creates a [GitHub release] in draft mode. Creating a `draft-release` for _every_ nightly build (everyday so) is problematic because we might be overwhelmed by the number of draft releases.

We could [override the assets] of the previous nightly release (instead of recreating a new one). The problem with this approach is that GitHub sort the releases by creation date: nightly release will be put behind newer releases.

So the idea is to _remove_ the previous nigthly release and recreate a new one. In order to do so, the API to [delete a release] requires a `release_id`.

[override the assets]: https://stackoverflow.com/questions/62934246/github-update-overwrite-existing-asset-of-a-release
[GitHub release]: https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases
```

[trigger a workflow on a schedule]: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule

> [!CAUTION]
> Event triggered by a GitHub actions don't trigger a workflow. So everything should be done in a single workflow.

### Slack notification (optional)

We could send a Slack notification when a nightly build finished (successfully or not)
With the link to download the generated build.

Slack provide GitHub action for that: <https://github.com/slackapi/slack-github-action>

### Improve the unique dev format

Currently, `releaser.py` generate a dev version by appending the `dev` part at the end of the current version before the local part.

The dev part is formatted that way `dev.{number}`. The `{number}` value is calculated from the current date with the following formula:

$f(year, month, day) = year * 366 + month * 31 + day$

> [!NOTE]
> If you are careful, the value $372$ ($12 * 31 = 372$) should have been used instead of $366$ (that's a mistake)

For the date `2024-02-27 15:50:28+01:00` that gives `dev.740873`:

$$f(2024, 2, 27) = 2024 * 366 + 2 * 31 + 27$$
$$f(2024, 2, 27) = 740873$$

> The formula could be changed to:
> 
> - A _timestamp_: For `2024-02-27 15:50:28+01:00` we get `dev.1709045428`
> 
> - A _timestamp_ without the time info:
> 
>   For `2024-02-27 15:50:28+01:00`, the value is calculated with:
> 
>   $$S = 24 * 3600$$
>   $$S = 86400$$
>   $$timestamp = 1709045428$$
>   $$g(timestamp) = floor(timestamp / S)$$
>   $$g(1709045428) = 19780$$
> 
>   That gives `dev.19780`, converted back to a timestamp $19780 * S = 1708992000$, that give `2024-02-27 01:00:00+01:00`
> 
> - The corrected formulas using $372$:
> 
>   For `2024-02-27 15:50:28+01:00`, the value is calculated with:
> 
>   $$f'(year, month, day) = year * 372 + month * 31 + day$$
>   $$f'(2024, 2, 27) = 753017$$
> 
>   That would give `dev.753017`
> 
> - A date (`dev.YYYYMMDD`): For `2024-02-27 15:50:28+01:00` That gives `dev.20240227`

Reviewer: ```suggestion
Considering the previous example date, the formula could be changed to:

- A _timestamp_: `dev.1709045428`
- A _timestamp_ without the time info ($floor(timestamp / 24 * 3600)$ : `dev.19780` 
- The current formula with fixed value ($year * 372 + month * 31 + day$): `dev.753017`
- A date format (`dev.YYYYMMDD`): `dev.20240227`
```
